File: src/middleware/services/teacherService.js
const db = require('../../../config/db');

function getAllTeachers() {
    return db.prepare("SELECT * FROM teachers ORDER BY teacher_id DESC").all();
}

function getTeacherById(id) {
    return db.prepare("SELECT * FROM teachers WHERE teacher_id = ?").get(id);
}

function checkTeacherLogin(loginIdentifier, password) {
    // Matches by email OR employee_id
    const teacher = db.prepare(`
        SELECT * FROM teachers
        WHERE (email = ? OR employee_id = ?) AND password = ?
    `).get(loginIdentifier, loginIdentifier, password);
    return teacher;
}

function checkEmployeeIdExists(employeeId, excludeId = null) {
    if (excludeId) {
        const teacher = db.prepare(
            "SELECT teacher_id FROM teachers WHERE employee_id = ? AND teacher_id != ?"
        ).get(employeeId, excludeId);
        return !!teacher;
    } else {
        const teacher = db.prepare(
            "SELECT teacher_id FROM teachers WHERE employee_id = ?"
        ).get(employeeId);
        return !!teacher;
    }
}

function insertTeacher(data) {
    const { employee_id, name, gender, department, designation, employment_type, email, phone, password } = data;
    const pwd = password && password.trim() !== '' ? password : employee_id;
    const stmt = db.prepare(`
        INSERT INTO teachers (employee_id, name, gender, department, designation, employment_type, email, phone, password)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);
    const info = stmt.run(employee_id, name, gender, department, designation, employment_type, email, phone, pwd);
    return info.lastInsertRowid;
}

function updateTeacher(id, data) {
    const { employee_id, name, gender, department, designation, employment_type, email, phone, password } = data;
    if (password && password.trim() !== '') {
        const stmt = db.prepare(`
            UPDATE teachers
            SET employee_id = ?, name = ?, gender = ?, department = ?, designation = ?, employment_type = ?, email = ?, phone = ?, password = ?
            WHERE teacher_id = ?
        `);
        return stmt.run(employee_id, name, gender, department, designation, employment_type, email, phone, password, id);
    } else {
        const stmt = db.prepare(`
            UPDATE teachers
            SET employee_id = ?, name = ?, gender = ?, department = ?, designation = ?, employment_type = ?, email = ?, phone = ?
            WHERE teacher_id = ?
        `);
        return stmt.run(employee_id, name, gender, department, designation, employment_type, email, phone, id);
    }
}

function deleteTeacher(id) {
    return db.prepare("DELETE FROM teachers WHERE teacher_id = ?").run(id);
}

function getTeacherCount() {
    return db.prepare("SELECT COUNT(*) as count FROM teachers").get().count;
}

module.exports = {
    getAllTeachers,
    getTeacherById,
    checkTeacherLogin,
    checkEmployeeIdExists,
    insertTeacher,
    updateTeacher,
    deleteTeacher,
    getTeacherCount
};

File: app/services/teacher_service.py
from database.database import get_connection


def get_all_teachers():

    connection = get_connection()

    cursor = connection.cursor()

    cursor.execute(
        """
        SELECT *
        FROM teachers
        ORDER BY teacher_id DESC
        """
    )

    teachers = cursor.fetchall()

    connection.close()

    return teachers


def check_employee_id_exists(employee_id):

    connection = get_connection()

    cursor = connection.cursor()

    cursor.execute(
        """
        SELECT teacher_id
        FROM teachers
        WHERE employee_id = ?
        """,
        (employee_id,)
    )

    teacher = cursor.fetchone()

    connection.close()

    return teacher is not None


def insert_teacher(
    employee_id,
    name,
    gender,
    department,
    designation,
    employment_type,
    email,
    phone
):

    connection = get_connection()

    cursor = connection.cursor()

    cursor.execute(
        """
        INSERT INTO teachers
        (
            employee_id,
            name,
            gender,
            department,
            designation,
            employment_type,
            email,
            phone
        )

        VALUES
        (
            ?, ?, ?, ?, ?, ?, ?, ?
        )
        """,
        (
            employee_id,
            name,
            gender,
            department,
            designation,
            employment_type,
            email,
            phone
        )
    )

    connection.commit()

    connection.close()


def update_teacher(
    teacher_id,
    employee_id,
    name,
    gender,
    department,
    designation,
    employment_type,
    email,
    phone
):

    conn = get_connection()

    cursor = conn.cursor()

    cursor.execute("""
        UPDATE teachers
        SET
            employee_id = ?,
            name = ?,
            gender = ?,
            department = ?,
            designation = ?,
            employment_type = ?,
            email = ?,
            phone = ?
        WHERE teacher_id = ?
    """, (
        employee_id,
        name,
        gender,
        department,
        designation,
        employment_type,
        email,
        phone,
        teacher_id
    ))

    conn.commit()
    conn.close()


def delete_teacher(teacher_id):

    connection = get_connection()

    cursor = connection.cursor()

    cursor.execute(
        """
        DELETE FROM teachers
        WHERE teacher_id = ?
        """,
        (teacher_id,)
    )

    connection.commit()

    connection.close()
